sdk, programs/bpf_loader: add sol_remaining_compute_units syscall

ckamm opened the pull request:
#### Problem

Instructions may want to maximise their use of the compute budget, without risking going over budget. To do so, they ought to be able to inspect the current budget consumed.

#### Summary of Changes

Implement a syscall that allows SBF programs to make queries of `sol_remaining_compute_units` at arbitrary points in execution.

- Fixes #18134
- Rebased revival of #18227 (note that I dropped a - it seemed to me - unrelated change to the cost of `sol_log_compute_units`)

#### TODO

- [x] Feature gate pubkey
- [x] Docs link to "invalid_link". The existing link for another syscall goes to an arbitrary blob on github. I don't think I can fill that before the PR is merged.
- [x] The C sdk header additions are untested. How is this typically tested?
- [x] `programs/sbf/rust/early_terminate` has a program_test, so I could check the call works as expected. Is this the usual way to test?

ckamm commented:
@KirillLykov This is the rebased patch we talked about in #18227

KirillLykov commented:
To me looks like a useful feature but let's invite @t-nelson who was discussing it with @jon-chuang in the original PR

t-nelson commented:
is there a simd to go along with this?  i don't recall any particularly compelling use cases for such a feature

ckamm commented:
@t-nelson There's no simd. If one is needed, I can make one.

I pointed at the usecases for mango here https://github.com/solana-labs/solana/pull/18227#issuecomment-1543652969: Instructions that can process a variable number of things (events in cranks, orderbook matching depth) would like to know how much compute is remaining so they can make best use of requested CU.

mvines commented:
Yes SIMD first please. We now aim to route all consensus-affecting changes through that design and socialization process before an implementation is added to any validator client

t-nelson commented:
> @t-nelson There's no simd. If one is needed, I can make one.

yes please!

> I pointed at the usecases for mango here [#18227 (comment)](https://github.com/solana-labs/solana/pull/18227#issuecomment-1543652969): Instructions that can process a variable number of things (events in cranks, orderbook matching depth) would like to know how much compute is remaining so they can make best use of requested CU.

yeah but your can probably tell that i wasn't particularly convinced by my comments in that conversation :wink:

ckamm commented:
SIMD: https://github.com/solana-foundation/solana-improvement-documents/pull/49

mvines commented on sdk/program/src/program_stubs.rs:
```diff
@@ -30,6 +30,10 @@ pub trait SyscallStubs: Sync + Send {
     }
+    fn sol_remaining_compute_units(&self) -> u64 {
+        sol_log("SyscallStubs: sol_remaining_compute_units() defaulting to u64::MAX");
+        u64::MAX
```
total nit: 0 instead of `u64::MAX` feels like a better stub value here, as that'd cause any programs under test that are looping on the syscall to break instead of infinite looping.

ckamm replied:
Changed

mvines commented on sdk/src/feature_set.rs:
```diff
@@ -665,6 +665,11 @@ pub mod checked_arithmetic_in_fee_validation {
 }
 
+pub mod remaining_compute_units_syscall_enabled {
+    // TODO: replace by team
+    solana_sdk::declare_id!("GH4auHBrmun5EYthnRw8sQ6JZhGWGsUJNPzuNGKEq37C");
```
Here's the pubkey for a keypair that I hold
```suggestion
    solana_sdk::declare_id!("5TuppMutoyzhUSfuYdhgzD47F92GL1g89KpCZQKqedxP");
```

mvines commented:
> SIMD: [solana-foundation/solana-improvement-documents#49](https://github.com/solana-foundation/solana-improvement-documents/pull/49)

SIMD is merged! This PR isn't blocked anymore 🎊

t-nelson commented on programs/bpf_loader/src/syscalls/mod.rs:
```diff
@@ -1791,6 +1801,26 @@ declare_syscall!(
+        _memory_mapping: &mut MemoryMapping,
+    ) -> Result<u64, Error> {
+        let budget = invoke_context.get_compute_budget();
+        consume_compute_meter(invoke_context, budget.syscall_base_cost)?;
```
we should probably go ahead and give this its own cost field for maintainability's sake

ckamm replied:
I called this `ComputeBudget::get_remaining_compute_units_cost` and made it 100 - sound good?

t-nelson replied:
seem's like we didn't deem the base cost worthy of a constant for w/e reason, so `100` works for me

t-nelson commented on programs/bpf_loader/src/syscalls/mod.rs:
```diff
@@ -1791,6 +1801,26 @@ declare_syscall!(
+        consume_compute_meter(invoke_context, budget.syscall_base_cost)?;
+
+        use solana_rbpf::vm::ContextObject;
+        Ok(invoke_context.get_remaining())
```
wow. we dropped the ball on that function name :facepalm:

ckamm commented:
Open questions for this PR:
- The C sdk header additions are untested. How is this typically tested?
- programs/sbf/rust/early_terminate has a program_test, so I could check the call works as expected. Is this the usual way to test?
- shall I merge in master to resolve the conflicts, or do you typically squash+rebase?

t-nelson commented:
>     * The C sdk header additions are untested. How is this typically tested?

most of the C tests are under `programs/sbf/c/` admittedly not great coverage there
 
>     * programs/sbf/rust/early_terminate has a program_test, so I could check the call works as expected. Is this the usual way to test?

looks like an ok test functionally.  i'd probably name it to better reflect the goal of testing `sol_remaining_compute_units()` though.  porting it to C would probably sufficiently resolve the above

>     * shall I merge in master to resolve the conflicts, or do you typically squash+rebase?

rebase for sure to appease ci. squash is up to you, gh will do it when we press the green button either way

ckamm commented:
Ok, I renamed "early_terminate" to "remaining_compute_units" and added a C version. Looks like the makefile in programs/sbf/c/makefile is only compiling these files though, there's no execution: is that all there is to it right now, or is there some script to deploy and call these?

Conflicts: I've merged in master for now since that's easier for review but will squash it away before merge.

t-nelson commented:
thanks! lgtm now. i'm going to tag in runtime folks for a quick sanity check

t-nelson commented:
@ckamm i think we're good to go here.  just wanted to confirm with you that the unchecked TODO items in the topic are in fact complete before hitting the green button

ckamm commented:
@t-nelson The thing that's still missing - and that the CI complains about - is the docs. The doc additions in developing-c/rust.md still have invalid links, because I don't know what the target link is going to be - I think that link can only be set once it's in the main solana repo?

t-nelson commented:
> @t-nelson The thing that's still missing - and that the CI complains about - is the docs. The doc additions in developing-c/rust.md still have invalid links, because I don't know what the target link is going to be - I think that link can only be set once it's in the main solana repo?

ah, weird. i must've been looking at a stale cached version of the page earlier... everything was green.

that repo link for `sol_log_compute_units()` is kinda bs, so far as "docs" go.  if you're intent on following that path, rather than adding actual documentation, i'd split the docs out and submit them in a followup pr.

ckamm commented:
@t-nelson I dropped the doc links - I agree they don't add much, but wanted to be consistent with what I saw on the other syscalls.

Also merged in master again to resolve a conflict. If CI passes, I think this could be merged. Feel free to squash.

KirillLykov commented:
@ckamm could you please re-resolve conflict again and when CI pass, ping @t-nelson?

t-nelson commented:
ah! apologies, i thought we'd already landed this one.  please ping me once rebased

ckamm commented:
Ok, I hope CI will just pass.

Note that I now saw the warning:
```
    // !!! ATTENTION !!!
    // When adding new features for RBPF here,
    // also add them to `Bank::apply_builtin_program_feature_transitions()`.
```
and I'm unclear on whether adding a syscall qualifies. Should I also add this feature there? The feature list there contains some syscalls, so probably yes?

KirillLykov commented:
@ckamm please rebase again

t-nelson reviewed (approved):
r+ ci

t-nelson commented:
> r+ ci

damn, jinxed it

```
programs/sbf/c/src/remaining_compute_units/remaining_compute_units.c:9: trailing whitespace.
```

KirillLykov commented:
don't want to be annoying but don't forget to rebase  @ckamm

ckamm commented:
@KirillLykov Ah, I thought resolving all merge conflicts with master would be sufficient, so you could just squash-apply it. I've squashed it and brought it up to date with master yet again.

t-nelson commented:
```
error: integer arithmetic detected
  --> rust/remaining_compute_units/src/lib.rs:23:9
   |
23 |         i += 1;
   |         ^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#integer_arithmetic
   = note: requested on the command line with `-D clippy::integer-arithmetic`
```

you probably want this guy, https://docs.rs/solana-sdk/latest/solana_sdk/macro.saturating_add_assign.html

might wanna just run `./ci/test-sanity.sh` and `./ci/test-checks.sh` locally to speed iteration

ckamm commented:
@t-nelson Updated and rebased yet again (why do I have to do this when there's no relevant conflicts?)

Both scripts pass locally. (dropped the check for grcov, sccache, wasm-pack from ci/test-checks.sh to make it less annoying for me)

t-nelson commented:
hmm this failing test passes locally :thinking:

t-nelson commented:
> hmm this failing test passes locally thinking

n/m. fuck macos. gimme a few

t-nelson commented:
> gimme a few

blocked by https://github.com/solana-labs/solana/pull/32159#issuecomment-1673957510

t-nelson commented:
> > gimme a few
> 
> blocked by [#32159 (comment)](https://github.com/solana-labs/solana/pull/32159#issuecomment-1673957510)

ok, master test should be fixed now. please rebase (hopefully! :crossed_fingers:) one last time @ckamm

ckamm commented:
@t-nelson rebased, can you start CI? ci/test-checks.sh passed locally

t-nelson commented:
@ckamm amazing how bad a luck we're having getting this thing in.  can you rebase on 297ffad7977ac42cf5d08abec62c181e7f3d31e1 or later?

ckamm commented:
@t-nelson Done!

So many of these are just simple insert-insert conflicts because new features get added at the end. Maybe it could be worthwhile to not add at the end but sort them alphabetically? Then most additions would be in different places and the number of conflicts would go down.
